File: otto-mobile/otto-mobile/OttoAnalytics.swift
import FirebaseAnalytics
import FirebaseCore
import FirebaseCrashlytics

enum OttoAnalytics {
    private static var isConfigured = false

    static func configure() {
        guard !isConfigured else { return }
        FirebaseApp.configure()
        #if DEBUG
        Analytics.setAnalyticsCollectionEnabled(true)
        Crashlytics.crashlytics().setCrashlyticsCollectionEnabled(false)
        #endif
        isConfigured = true
    }

    static func setUserID(_ userID: String) {
        guard !userID.isEmpty else { return }
        Analytics.setUserID(userID)
        Crashlytics.crashlytics().setUserID(userID)
    }

    static func clearUserID() {
        Analytics.setUserID(nil)
        Crashlytics.crashlytics().setUserID("")
    }

    static func logSignUpComplete() {
        Analytics.logEvent("sign_up_complete", parameters: nil)
    }

    static func logOnboardingNameComplete() {
        Analytics.logEvent("onboarding_name_complete", parameters: nil)
    }

    static func logLocationSharingEnabled() {
        Analytics.logEvent("location_sharing_enabled", parameters: ["source": "map"])
    }

    static func logLocationSharingDisabled(reason: String) {
        Analytics.logEvent("location_sharing_disabled", parameters: ["reason": reason])
    }

    static func logDriveStarted(kind: String, routeID: String? = nil) {
        var params: [String: Any] = ["kind": kind]
        if let routeID, !routeID.isEmpty {
            params["route_id"] = routeID
        }
        Analytics.logEvent("drive_started", parameters: params)
    }

    static func logDriveCompleted(kind: String, distanceMeters: Double) {
        let bucket = max(0, Int((distanceMeters / 100.0).rounded()) * 100)
        Analytics.logEvent(
            "drive_completed",
            parameters: [
                "kind": kind,
                "distance_meters": bucket,
            ]
        )
    }

    static func logEventCheckIn(eventID: String) {
        let trimmed = eventID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        Analytics.logEvent("event_check_in", parameters: ["event_id": trimmed])
    }

    static func logAutoCheckInGeofenceRegistered(regionCount: Int) {
        Analytics.logEvent(
            "auto_checkin_geofence_registered",
            parameters: ["region_count": max(0, regionCount)]
        )
    }

    static func logAutoCheckInGeofenceEntered(eventID: String) {
        let trimmed = eventID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        Analytics.logEvent("auto_checkin_geofence_entered", parameters: ["event_id": trimmed])
    }

    static func logAutoCheckInSkipped(eventID: String, reason: String, trigger: String) {
        let trimmed = eventID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        Analytics.logEvent(
            "auto_checkin_skipped",
            parameters: [
                "event_id": trimmed,
                "reason": reason,
                "trigger": trigger,
            ]
        )
    }

    static func logAutoCheckInSuccess(eventID: String, trigger: String) {
        let trimmed = eventID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        Analytics.logEvent(
            "auto_checkin_success",
            parameters: [
                "event_id": trimmed,
                "trigger": trigger,
            ]
        )
    }

    static func logSquadCreated() {
        Analytics.logEvent("squad_created", parameters: nil)
    }

    static func logSquadJoined(source: String) {
        Analytics.logEvent("squad_joined", parameters: ["source": source])
    }

    static func logChatMessageSent(channel: String, attachmentType: String) {
        Analytics.logEvent(
            "chat_message_sent",
            parameters: [
                "channel": channel,
                "attachment_type": attachmentType,
            ]
        )
    }

    static func logGarageCarAdded(hasPhoto: Bool) {
        Analytics.logEvent(
            "garage_car_added",
            parameters: ["has_photo": hasPhoto ? "true" : "false"]
        )
    }

    static func logRouteSaved() {
        Analytics.logEvent("route_saved", parameters: nil)
    }

    static func analyticsDriveKind(_ kind: DriveSessionKind) -> String {
        switch kind {
        case .quick, .live:
            return "quick"
        case .route:
            return "route"
        }
    }
}

File: otto-mobile/otto-android/app/src/main/java/to/ottomot/driftd/core/analytics/OttoAnalytics.kt
package to.ottomot.driftd.core.analytics

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import kotlin.math.roundToInt
import to.ottomot.driftd.BuildConfig
import to.ottomot.driftd.DriveSessionKind

object OttoAnalytics {
    private var firebaseAnalytics: FirebaseAnalytics? = null
    private var isConfigured = false

    fun configure(context: Context) {
        if (isConfigured) return
        val analytics = FirebaseAnalytics.getInstance(context.applicationContext)
        firebaseAnalytics = analytics
        isConfigured = true
        if (BuildConfig.DEBUG) {
            analytics.setAnalyticsCollectionEnabled(true)
            FirebaseCrashlytics.getInstance().isCrashlyticsCollectionEnabled = false
        }
    }

    fun setUserID(userId: String) {
        if (userId.isBlank()) return
        firebaseAnalytics?.setUserId(userId)
        FirebaseCrashlytics.getInstance().setUserId(userId)
    }

    fun clearUserID() {
        firebaseAnalytics?.setUserId(null)
        FirebaseCrashlytics.getInstance().setUserId("")
    }

    fun logSignUpComplete() {
        logEvent("sign_up_complete")
    }

    fun logOnboardingNameComplete() {
        logEvent("onboarding_name_complete")
    }

    fun logLocationSharingEnabled() {
        logEvent(
            "location_sharing_enabled",
            bundleOf("source" to "map"),
        )
    }

    fun logLocationSharingDisabled(reason: String) {
        logEvent(
            "location_sharing_disabled",
            bundleOf("reason" to reason),
        )
    }

    fun logDriveStarted(
        kind: String,
        routeId: String? = null,
    ) {
        val params =
            Bundle().apply {
                putString("kind", kind)
                routeId?.trim()?.takeIf { it.isNotEmpty() }?.let { putString("route_id", it) }
            }
        logEvent("drive_started", params)
    }

    fun logDriveCompleted(
        kind: String,
        distanceMeters: Double,
    ) {
        val bucket = ((distanceMeters / 100.0).roundToInt() * 100).coerceAtLeast(0)
        logEvent(
            "drive_completed",
            bundleOf(
                "kind" to kind,
                "distance_meters" to bucket.toLong(),
            ),
        )
    }

    fun logEventCheckIn(eventId: String) {
        val trimmed = eventId.trim()
        if (trimmed.isEmpty()) return
        logEvent(
            "event_check_in",
            bundleOf("event_id" to trimmed),
        )
    }

    fun logSquadCreated() {
        logEvent("squad_created")
    }

    fun logSquadJoined(source: String) {
        logEvent(
            "squad_joined",
            bundleOf("source" to source),
        )
    }

    fun logChatMessageSent(
        channel: String,
        attachmentType: String,
    ) {
        logEvent(
            "chat_message_sent",
            bundleOf(
                "channel" to channel,
                "attachment_type" to attachmentType,
            ),
        )
    }

    fun logGarageCarAdded(hasPhoto: Boolean) {
        logEvent(
            "garage_car_added",
            bundleOf("has_photo" to if (hasPhoto) "true" else "false"),
        )
    }

    fun logRouteSaved() {
        logEvent("route_saved")
    }

    fun logDriveStartedFromSession(
        kind: DriveSessionKind,
        routeId: String? = null,
    ) {
        logDriveStarted(analyticsDriveKind(kind), routeId)
    }

    fun logDriveCompletedFromSession(
        kind: DriveSessionKind,
        distanceMeters: Double,
    ) {
        logDriveCompleted(analyticsDriveKind(kind), distanceMeters)
    }

    private fun analyticsDriveKind(kind: DriveSessionKind): String =
        when (kind) {
            DriveSessionKind.QUICK -> "quick"
            DriveSessionKind.ROUTE -> "route"
            DriveSessionKind.LIVE -> "quick"
        }

    private fun logEvent(
        name: String,
        params: Bundle? = null,
    ) {
        firebaseAnalytics?.logEvent(name, params)
    }

    private fun bundleOf(vararg pairs: Pair<String, Any>): Bundle =
        Bundle(pairs.size).apply {
            pairs.forEach { (key, value) ->
                when (value) {
                    is String -> putString(key, value)
                    is Long -> putLong(key, value)
                    is Int -> putLong(key, value.toLong())
                    is Boolean -> putString(key, if (value) "true" else "false")
                    else -> putString(key, value.toString())
                }
            }
        }
}
